src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -76,16 +79,6 @@ bool isSafePrimaryDataKeyType(const IDataType & data_type)
             const auto & data_type_map = static_cast<const DataTypeMap &>(data_type);
             return isSafePrimaryDataKeyType(*data_type_map.getKeyType()) && isSafePrimaryDataKeyType(*data_type_map.getValueType());
         }
-        case TypeIndex::Variant:
-        {
-            const auto & data_type_variant = static_cast<const DataTypeVariant &>(data_type);
-            const auto & data_type_variant_elements = data_type_variant.getVariants();
-            for (const auto & data_type_variant_element : data_type_variant_elements)
-                if (!isSafePrimaryDataKeyType(*data_type_variant_element))
-                    return false;
-
-            return false;
```

Comment:
Have you removed this code because it always returns false? This "return false" statement seems as a bug to me, tbh

Reply:
Yes, this code always returned false. I don't know why it was written this way. But Varaint type as well as Dynamic should be always marked as not safe because we cannot use comparison between Fields from these columns